### DISCUSSION TITLE
Getting 401 and 403 when running dotnet restore

### DISCUSSION THREAD

**Discussion by radulfr96** [community/community#26974] [Actions]
<p>I am trying to use the basic .NET Core Github workflow as shown <a href="https://github.com/actions/starter-workflows/blob/master/ci/dotnet-core.yml" rel="nofollow noopener">here</a>, but I am having trouble with nuget. When the workflow gets to the Install Depedencies which performs a donet restore, it keeps getting 403s and 401 errors when running on my WebAPI repo <a href="https://github.com/MyLibrary-com/my-library-web-api" rel="nofollow noopener">here</a>. I have tried using, <span class="math">{{secrets.GITHUB\_TOKEN}},&amp;nbsp;</span>{{secrets.PERSONAL_ACCESS_TOKEN}} with repo and read package permissions and using the actual token instead of the secret reference.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/0/0c28edcf334f5efdd74bb41a9b62c9eb7c5b42ad.jpeg" data-download-href="/uploads/short-url/1Jzq0yQWWSkCk7PT7Nhjhy7iRuR.jpg?dl=1" title="Nuget401403.JPG" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/20983472/181101026-958aaf5c-7931-4fe2-8d10-0b4f6558ed34.jpg" alt="Nuget401403.JPG" data-base62-sha1="1Jzq0yQWWSkCk7PT7Nhjhy7iRuR" width="690" height="318" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/0/0c28edcf334f5efdd74bb41a9b62c9eb7c5b42ad_2_690x318.jpg, https://aws1.discourse-cdn.com/github/optimized/2X/0/0c28edcf334f5efdd74bb41a9b62c9eb7c5b42ad_2_1035x477.jpg 1.5x, https://aws1.discourse-cdn.com/github/original/2X/0/0c28edcf334f5efdd74bb41a9b62c9eb7c5b42ad.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/0/0c28edcf334f5efdd74bb41a9b62c9eb7c5b42ad_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Nuget401403.JPG</span><span class="informations">1378×637 221 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>How do I fix this?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/radulfr96">@radulfr96</a> ,</p>
<p>In the " <strong>nuget.config</strong>" file, I see you set the Credentials like as below:</p>
<pre><code>&lt;packageSourceCredentials&gt;
  &lt;github&gt;
    &lt;add key="Username" value="radulfr96" /&gt;
    &lt;add key="ClearTextPassword" value="${{secrets.GITHUB_TOKEN}}" /&gt;
  &lt;/github&gt;
&lt;/packageSourceCredentials&gt;
</code></pre>
<p>You should not directly call the secrets in the " <strong>nuget.config</strong>" file, you need to set the secrets as environment variables at first, then call the environment variables in the " <strong>nuget.config</strong>" file.</p>
<pre><code>## in the workflow file
jobs:
  build:
    runs-on: ubuntu-latest
    env:
      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      PERSONAL_ACCESS_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
    steps:
    . . .

## in the "nuget.config" file
&lt;add key="ClearTextPassword" value="${GITHUB_TOKEN}" /&gt;
OR
&lt;add key="ClearTextPassword" value="${PERSONAL_ACCESS_TOKEN}" /&gt;
</code></pre>
<p>Typically, you can directly use the secrets in the value of the <strong>with</strong> key.</p>
<p>For example:</p>
<pre><code>- name: Checkout code
  uses: actions/checkout@v2
  with:
    token: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
</code></pre>
<p>Other situations, such as in the command lines or in the config/settings files, you’d better set the secrets as environment variables at first.</p>
<p>More details, you can reference the docs about <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets" rel="nofollow noopener"><strong>Secrets</strong></a>.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/radulfr96">@radulfr96</a> ,</p>
<p>How are things going? Is my suggestion helpful to you?</p>
<p>Please try it, and any progress, feel free to tell me.</p>

**Comment by radulfr96** [comment]
<p><a class="mention" href="/brightran">@brightran</a> </p>
<p>I haven’t had the chance to try it yet. This is a hobby app I work on, on the weekends. I’ll update you then. But your solution looks promising. Thanks</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/radulfr96">@radulfr96</a> ,</p>
<p>Okay, any update, feel free to tell me.</p>

**Comment by radulfr96** [comment]
<p><a class="mention" href="/brightran">@brightran</a> </p>
<p>I made the changes you reccomended and I am still getting 401s. Do you have any other sugesstions?</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/radulfr96">@radulfr96</a> ,</p>
<p>How about " <strong>&lt;add key=“ClearTextPassword” value="%GITHUB_TOKEN%" /&gt;</strong>"?</p>
<p>I tested, it can work fine, see the example here:</p>
<p><a href="https://github.com/TestGitHubPackagesRML/TestNuGet/blob/master/TestNuGet/nuget.config" rel="nofollow noopener">https://github.com/TestGitHubPackagesRML/TestNuGet/blob/master/TestNuGet/nuget.config</a></p>
<p><a href="https://github.com/TestGitHubPackagesRML/TestNuGet/blob/master/TestNuGet/TestNuGet.csproj" rel="nofollow noopener">https://github.com/TestGitHubPackagesRML/TestNuGet/blob/master/TestNuGet/TestNuGet.csproj</a></p>
<p><a href="https://github.com/TestGitHubPackagesRML/TestNuGet/blob/master/.github/workflows/CI.yml" rel="nofollow noopener">https://github.com/TestGitHubPackagesRML/TestNuGet/blob/master/.github/workflows/CI.yml</a></p>

**Comment by radulfr96** [comment]
<p><a class="mention" href="/brightran">@brightran</a>,</p>
<p>Hi </p>
<p>I changed it to that and it appears to work but now I am getting </p>
<p>error NU1101: Unable to find package Microsoft.EntityFrameworkCore. No packages exist with this id in source(s): github</p>
<p>Which I assume is because my nuget.config does not reference <a href="http://nuget.org" rel="nofollow noopener">nuget.org</a></p>
<p>Edit: I fixed it and it is now building, thank you.</p>
